From 3ec771fdd9832f792b00e452482c6745cc1cfd21 Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Fri, 5 May 2023 16:11:10 +0800
Subject: [PATCH 1630/1646] [Huawei] mm/dpool: Show dynamic pool info during
 OOM

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188703

--------------------------------

Show dynamic pool info during OOM.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 include/linux/dynamic_hugetlb.h |  2 ++
 lib/show_mem.c                  |  3 +++
 mm/dynamic_hugetlb.c            | 16 ++++++++++++++++
 3 files changed, 21 insertions(+)

diff --git a/include/linux/dynamic_hugetlb.h b/include/linux/dynamic_hugetlb.h
index 48b8bd3821bd..31f81afa6d88 100644
--- a/include/linux/dynamic_hugetlb.h
+++ b/include/linux/dynamic_hugetlb.h
@@ -136,6 +136,7 @@ bool page_belong_to_dynamic_hugetlb(struct page *page);
 
 int dpool_init(struct dpool_info *arg);
 void dpool_report_meminfo(struct seq_file *m);
+void dpool_show_mem_info(void);
 
 #else
 
@@ -213,6 +214,7 @@ bool page_belong_to_dynamic_hugetlb(struct page *page)
 struct dpool_info {};
 static inline int dpool_init(struct dpool_info *arg) { return 0; }
 static inline void dpool_report_meminfo(struct seq_file *m) {}
+static inline void dpool_show_mem_info(void) {}
 
 #endif /* CONFIG_DYNAMIC_HUGETLB */
 #endif /* __LINUX_DYNAMIC_HUGETLB_H */
diff --git a/lib/show_mem.c b/lib/show_mem.c
index 11751aebc98f..55e754cd7ef1 100644
--- a/lib/show_mem.c
+++ b/lib/show_mem.c
@@ -7,6 +7,7 @@
 
 #include <linux/mm.h>
 #include <linux/cma.h>
+#include <linux/dynamic_hugetlb.h>
 
 void show_mem(unsigned int filter, nodemask_t *nodemask)
 {
@@ -42,4 +43,6 @@ void show_mem(unsigned int filter, nodemask_t *nodemask)
 	printk("%lu pages hwpoisoned\n", atomic_long_read(&num_poisoned_pages));
 #endif
 	reliable_show_mem_info();
+
+	dpool_show_mem_info();
 }
diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index 5e9ae0f45019..e72d9a0812f5 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -1456,3 +1456,19 @@ void dpool_report_meminfo(struct seq_file *m)
 		   (free_pages + used_pages) << (PAGE_SHIFT - 10),
 		   free_pages << (PAGE_SHIFT - 10));
 }
+
+void dpool_show_mem_info(void)
+{
+	unsigned long free_pages = 0, used_pages = 0;
+	struct dhugetlb_pool *hpool = dpool_global_hpool;
+
+	if (!dpool_enabled)
+		return;
+
+	if (hpool)
+		hugetlb_sum_normal_pages(hpool, &free_pages, &used_pages);
+
+	pr_info("DPoolTotal: %lu kB\n", (free_pages + used_pages)
+						<< (PAGE_SHIFT - 10));
+	pr_info("DPoolFree: %lu kB\n", free_pages << (PAGE_SHIFT - 10));
+}
-- 
2.17.1

